### DISCUSSION TITLE
BTree Unit Tests

### DISCUSSION THREAD

**Discussion by RahulKushwaha** [sunbains/embedded-innodb#64] [General]
I started writing unit tests for BTree in an attempt to understand it better and hopefully get rid of the index lock. After multiple hacks I was able to get `optimistic_insert` to return `DB_SUCCESS` (yay!!). [Code Sample](https://github.com/RahulKushwaha/embedded-innodb/pull/10). 
I see that all the things like(Trx, locks, rollback seg, etc) are deeply integrated in the btree code. I guess not the modification of these components but validations. Validations like (index is locked, etc). And this makes unit-testing a little hard(as I need to properly setup things, and eventually this looks not a unit-test :D ) .  

I was wondering if we would like to keep this deep integration as a long term goal? On the other hand I feel untangling this deep integration will be hard.

**Comment by sunbains** [comment]
One way to look at this is that the validation is probably required due to the absence of unit tests. If the unit tests can guarantee similar correctness then the usefulness of this validation is moot. However, I doubt that unit tests can guarantee that, one other approach could be to pass context in that does these checks or any arbitrary tests.

**Reply by RahulKushwaha** [reply]
Makes sense. Let me write a couple of more scenarios, and I will have a better grasp on deep integration, and what the shape of context will look like.

**Reply by filasieno** [reply]
Question: what is your feeling on adding a standard UnitTesting framework. 
Develop an ad-hoc solution ? 
Reuse Catch? Google Test ? CTest ? ...

**Reply by sunbains** [reply]
My personal preference is to use something off the shelf. Otherwise it will be a big distraction.
